kernel/mem_domain.c

```diff
@@ -0,0 +1,294 @@
+			if (!parts[i]) {
+				irq_unlock(key);
+				return -EINVAL;
+			}
+			domain->partitions[i] = *parts[i];
+		}
+	}
+
+	sys_dlist_init(&domain->mem_domain_q);
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+int k_mem_domain_destroy(struct k_mem_domain *domain)
+{
+	unsigned int key;
+	sys_dnode_t *node, *next_node;
+
+	if (!domain) {
+		return -EINVAL;
+	}
+
+	key = irq_lock();
+
+	SYS_DLIST_FOR_EACH_NODE_SAFE(&domain->mem_domain_q, node, next_node) {
+		struct k_thread *thread =
+			CONTAINER_OF(node, struct k_thread, mem_domain_info);
+
+		/*
+		 * Call arch specific API to disable thread's memory
+		 * partitions.
+		 */
+		arch_disable_all_mem_partitions(thread);
+
+		sys_dlist_remove(&thread->mem_domain_info.mem_domain_q_node);
+		thread->mem_domain_info.mem_domain = NULL;
+	}
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+int k_mem_domain_add_partition(struct k_mem_domain *domain,
+			       struct k_mem_partition *part)
+{
+	int p_idx;
+	unsigned int key;
+	sys_dnode_t *node, *next_node;
+
+	if (!domain || !part) {
+		return -EINVAL;
+	}
+
+	if (part->start + part->size <= part->start) {
+		return -EINVAL;
+	}
+
+	key = irq_lock();
+
+	for (p_idx = 0; p_idx < max_partitions; p_idx++) {
+		/* A zero-sized partition denotes it's a free partition */
+		if (domain->partitions[p_idx].size == 0) {
+			break;
+		}
+	}
+
+	/* No free partition */
+	if (p_idx == max_partitions) {
+		irq_unlock(key);
+		return -ENOMEM;
+	}
+
+	domain->partitions[p_idx].start = part->start;
+	domain->partitions[p_idx].size = part->size;
+	domain->partitions[p_idx].attr = part->attr;
+
+	domain->num_partitions++;
+
+	SYS_DLIST_FOR_EACH_NODE_SAFE(&domain->mem_domain_q, node, next_node) {
+		struct k_thread *thread =
+			CONTAINER_OF(node, struct k_thread, mem_domain_info);
+
+		/*
+		 * Call arch specific API to configure a single
+		 * memory partition.
+		 */
+		arch_configure_mem_partition(thread, p_idx);
+	}
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+int k_mem_domain_remove_partition(struct k_mem_domain *domain,
+				  struct k_mem_partition *part)
+{
+	int p_idx;
+	unsigned int key;
+	sys_dnode_t *node, *next_node;
+
+	if (!domain || !part) {
+		return -EINVAL;
+	}
+
+	key = irq_lock();
+
+	/* find a partition that matches the given start and size */
+	for (p_idx = 0; p_idx < max_partitions; p_idx++) {
+		if (domain->partitions[p_idx].start == part->start &&
+		    domain->partitions[p_idx].size == part->size) {
+			break;
+		}
+	}
+
+	/* Not found */
+	if (p_idx == max_partitions) {
+		irq_unlock(key);
+		return -EINVAL;
+	}
+
+	SYS_DLIST_FOR_EACH_NODE_SAFE(&domain->mem_domain_q, node, next_node) {
+		struct k_thread *thread =
+			CONTAINER_OF(node, struct k_thread, mem_domain_info);
+
+		/*
+		 * Call arch specific API to disable a single
+		 * memory partition.
+		 */
+		arch_disable_mem_partition(thread, p_idx);
+	}
+
+	domain->partitions[p_idx].start = 0;
+	domain->partitions[p_idx].size = 0;
+	domain->partitions[p_idx].attr = 0;
+
+	domain->num_partitions--;
```

Comment:
This is not going to work.
What you are doing here is essentially deleting the *last* partition. If p_idx is not the last partition, this logic is incorrect.

Reply:
The logic here should be fine.
I've modified my sample to delete non-last partition and it worked fine.
(must fix the bug I mention below first)
But I found a bug in arm_core_mpu_configure_mem_domain() would disable wrong MPU regions after deleting non-last partition.